### PR TITLE
Measure cost of context switch in the firmware

### PR DESCRIPTION
**Pull request by francois141**

### CODE REVIEW

**Review by CharlyCst — COMMENTED** [review]
Awesome, thanks! Here are a few comments in the review.

In particular I am not sure about the use of both `mcycle` and `cycle`, is there a reason for that?
And also we should not introduce assembly except in the `arch` module, see comment for the details.

File: firmware/csr_ops/perf_counters.rs
[the remark this answers is not in the document]
Comment (francois141) [reply firmware/csr_ops/perf_counters.rs:1]: Not really as I emulate mcycle :(

File: firmware/tracing_firmware/main.rs [hunk 1]
```diff
@@ -15,11 +15,88 @@ fn enable_mcycle_in_smode() {
...
+macro_rules! trigger_ctx_switch {
+    ($register:expr) => {{
+        let begin: u64;
+        let end: u64;
+
+        unsafe {
+            // Read the first specified register
+            read_register!($register, begin);
+            // We can trigger any illegal instruction - the handler doesn't do anything
+            asm!("csrr {}, mcycle", out(reg) _);
```
[the remark this answers is not in the document]
Comment (francois141) [reply firmware/tracing_firmware/main.rs:37]: Actually mcycle causes a trap but no cycle in s-mode

File: firmware/tracing_firmware/main.rs [hunk 2]
```diff
@@ -15,11 +15,88 @@ fn enable_mcycle_in_smode() {
...
+    unsafe {
+        asm!(
+        "csrw mtvec, {mtvec}", // Write mtvec with trap handler
+        mtvec = in(reg) trap,
+        );
+    }
+
+    log::info!("Start benchmarking from Firmware");
+
+    benchmark!("mcycle");
```
[the remark this answers is not in the document]
Comment (francois141) [reply firmware/tracing_firmware/main.rs:94]: Cycle doesn't work in U-mode

File: src/virt.rs [hunk 3]
```diff
@@ -968,7 +968,20 @@ impl RegisterContextGetter<Csr> for VirtContext {
...
+                #[cfg(feature = "userspace")]
+                return 0;
+
+                #[cfg(not(feature = "userspace"))]
+                unsafe {
+                    use core::arch::asm;
+                    let mcycle_value: usize;
+                    asm!("csrr {}, mcycle", out(reg) mcycle_value);
+                    mcycle_value
+                }
```
[the remark this answers is not in the document]
Comment (francois141) [reply src/virt.rs:983]: Good to know thanks!!!!

**Comment by francois141** [comment]
@CharlyCst What I wanted to do is unfortunately not possible, so I think I will close the issue :(

**Pull request closed without merging**